src/main/java/seedu/bookmark/logic/parser/FindCommandParser.java

```diff
@@ -1,33 +1,96 @@
-        String trimmedArgs = args.trim();
-        if (trimmedArgs.isEmpty()) {
+        ArgumentMultimap argMultimap =
+                ArgumentTokenizer.tokenize(args,
+                        PREFIX_NAME, PREFIX_GENRE, PREFIX_COMPLETED, PREFIX_NOT_COMPLETED, PREFIX_TAG);
+
+        //checking if user had input more than one prefix. put the input prefix in inputPrefix.
+        if (argMultimap.getValue(PREFIX_NAME).isPresent()) {
+            prefixCount += 1;
+            inputPrefix = PREFIX_NAME;
+        }
+        if (argMultimap.getValue(PREFIX_GENRE).isPresent()) {
+            prefixCount += 1;
+            inputPrefix = PREFIX_GENRE;
+        }
+        if (argMultimap.getValue(PREFIX_COMPLETED).isPresent()) {
+            prefixCount += 1;
+            inputPrefix = PREFIX_COMPLETED;
+        }
+        if (argMultimap.getValue(PREFIX_NOT_COMPLETED).isPresent()) {
+            prefixCount += 1;
+            inputPrefix = PREFIX_NOT_COMPLETED;
+        }
+        if (argMultimap.getValue(PREFIX_TAG).isPresent()) {
+            prefixCount += 1;
+            inputPrefix = PREFIX_TAG;
+        }
+        if (prefixCount != 1) { //if more than 1 input prefix, we throw an error.
             throw new ParseException(
                     String.format(MESSAGE_INVALID_COMMAND_FORMAT, FindCommand.MESSAGE_USAGE));
         }
+        prefixCount = 0; //set count back to 0
 
-        String[] nameKeywords = trimmedArgs.split("\\s+");
+        String trimmedArgs = argMultimap.getValue(inputPrefix).get().trim();
+        if (trimmedArgs.isEmpty() && inputPrefix != PREFIX_COMPLETED && inputPrefix != PREFIX_NOT_COMPLETED) {
+            throw new ParseException(
+                    String.format(MESSAGE_INVALID_COMMAND_FORMAT, FindCommand.MESSAGE_USAGE));
+        }
 
-        return new FindCommand(new NameContainsKeywordsPredicate(Arrays.asList(nameKeywords)));
-    }
+        if (!trimmedArgs.isEmpty() && (inputPrefix == PREFIX_COMPLETED || inputPrefix == PREFIX_NOT_COMPLETED)) {
+            throw new ParseException(
+                    String.format(MESSAGE_INVALID_COMMAND_FORMAT, FindCommand.MESSAGE_USAGE));
+        }
+
+        String[] keywords = trimmedArgs.split("\\s+");
 
+        Predicate<Book> predicate = null;
+
+        if (inputPrefix == PREFIX_NAME) {
+            predicate = new NameContainsKeywordsPredicate(Arrays.asList(keywords));
+        } else if (inputPrefix == PREFIX_GENRE) {
+            predicate = new GenreContainsKeywordsPredicate(Arrays.asList(keywords));
+        } else if (inputPrefix == PREFIX_COMPLETED) {
+            predicate = new BookCompletedPredicate();
+        } else if (inputPrefix == PREFIX_NOT_COMPLETED) {
+            predicate = new BookNotCompletedPredicate();
+        } else if (inputPrefix == PREFIX_TAG) {
+            predicate = new TagContainsKeywordsPredicate(Arrays.asList(keywords));
+        }
+
+        return new FindCommand(predicate);
+    }
```

Comment:
This method looks quite long (60 LOC) but the recommended SLAP length is ~30
Maybe it will be better to separate the checking of Valid Find command and the parsing of the FindCommand?

Reply:
That's a good suggestion Penn Han, will work on it now!

Reply:
I have made the changes you suggested Penn Han, do let me know if it looks good to you. Thanks!